### PR TITLE
chore(planner): Refactor data type of `ScalarExpr`

### PR DESCRIPTION
**Pull request by leiysky**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Get data type of `ScalarExpr` with type checking to prevent unmatched data type.

### CODE REVIEW

**Review by BohuTANG — APPROVED** [review]

File: src/query/sql/src/planner/plans/scalar_expr.rs [hunk 1]
```diff
@@ -43,20 +43,24 @@ pub enum ScalarExpr {
...
-            ScalarExpr::OrExpr(scalar) => (*scalar.return_type).clone(),
-            ScalarExpr::NotExpr(scalar) => (*scalar.return_type).clone(),
-            ScalarExpr::ComparisonExpr(scalar) => (*scalar.return_type).clone(),
-            ScalarExpr::AggregateFunction(scalar) => (*scalar.return_type).clone(),
-            ScalarExpr::FunctionCall(scalar) => (*scalar.return_type).clone(),
-            ScalarExpr::CastExpr(scalar) => (*scalar.target_type).clone(),
-            ScalarExpr::SubqueryExpr(scalar) => scalar.data_type(),
-            ScalarExpr::Unnest(scalar) => (*scalar.return_type).clone(),
-        }
+    // pub fn data_type(&self) -> DataType {
```
Comment (xudong963) [inline src/query/sql/src/planner/plans/scalar_expr.rs:46]: useless?

**Review by xudong963 — APPROVED** [review]

**Comment by BohuTANG** [comment]
```
08_00_01_complex:                                                       [ FAIL ] - return code 1
, result:

1	(1,'a')
2	(3,'b')
3	(3,'c')
1	(1,'a')
1	(1,'a')
3	(3,'c')
2	(3,'b')
3	(3,'c')
b
c
a
b
c
(1,'a')
(3,'b')
(3,'c')
1	1	a	(1,'a')
2	3	b	(3,'b')
3	3	c	(3,'c')
3	3	c	(3,'c')
1	1	a	(1,'a')
2	3	b	(3,'b')
1	(1,'a')
2	(3,'b')
1
3
1
2
NULL
(NULL,NULL,NULL,NULL,NULL,['http://hl7.org/fhir/us/core/StructureDefinition/us-core-patient'],NULL,NULL)
(NULL,NULL,NULL,NULL,NULL,['http://hl7.org/fhir/us/core/StructureDefinition/us-core-patient'],NULL,NULL)
ERROR 1105 (HY000) at line 1: Code: 1002, displayText = Unsupported column type:Nullable(Nullable(Array(Nullable(String)))).
[(NULL,NULL,'official','Jennifer Infection Control','Infection Control',['Jennifer'],NULL,NULL,NULL),(NULL,NULL,'usual','Jennifer Infection Control','Infection Control',['Jennifer'],NULL,NULL,NULL)]
[(NULL,NULL,'official',NULL,'Quigley282',['Wilmer32'],['Mr.'],NULL,NULL)]
[(NULL,NULL,'official',NULL,'Collier206',['Towanda270'],['Mrs.'],NULL,NULL),(NULL,NULL,'maiden',NULL,'Mayer370',['Towanda270'],['Mrs.'],NULL,NULL)]
(NULL,NULL,'official','Jennifer Infection Control','Infection Control',['Jennifer'],NULL,NULL,NULL)
(NULL,NULL,'official',NULL,'Quigley282',['Wilmer32'],['Mr.'],NULL,NULL)
(NULL,NULL,'official',NULL,'Collier206',['Towanda270'],['Mrs.'],NULL,NULL)
ERROR 1105 (HY000) at line 1: Code: 1002, displayText = Unsupported column type:Nullable(Nullable(String)).
(NULL,NULL,'usual','Jennifer Infection Control','Infection Control',['Jennifer'],NULL,NULL,NULL)
NULL
(NULL,NULL,'maiden',NULL,'Mayer370',['Towanda270'],['Mrs.'],NULL,NULL)
ERROR 1105 (HY000) at line 1: Code: 1002, displayText = Unsupported column type:Nullable(Nullable(Array(Nullable(String)))).
```

**Review by sundy-li — APPROVED** [review]

**Comment by leiysky** [comment]
This PR is blocked by #10468

**Comment by BohuTANG** [comment]
@mergify update

**Pull request merged**